storage/scratchpad/benches/sparse_merkle.rs

```diff
@@ -5,80 +5,237 @@ use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criteri
-                        .batch_update(one_large_batch, &proof_reader)
-                        .unwrap();
-                },
-                BatchSize::LargeInput,
-            )
-        });
+        let existing_state = zip_eq(&keys, &values)
+            .map(|(key, value)| (*key, value))
+            .collect::<Vec<_>>();
+        let mut naive_base_smt = NaiveSmt::new(existing_state.as_slice());
+
+        // group: insert to a committed SMT ("unknown" root)
+        let base_committed = Group {
+            name: "insert to committed base".into(),
+            blocks: block_sizes
+                .iter()
+                .map(|block_size| {
+                    let updates = Self::gen_updates(&mut rng, &keys, *block_size);
+                    let proof_reader = Self::gen_proof_reader(&mut naive_base_smt, &updates);
+                    Block {
+                        smt: SparseMerkleTree::new(naive_base_smt.get_root_hash()),
+                        updates,
+                        proof_reader,
+                    }
+                })
+                .collect(),
+        };
+
+        // group: insert to an uncommitted SMT (some structures in mem)
+        let base_uncommitted = Group {
+            name: "insert to uncommitted base".into(),
+            blocks: base_committed
+                .blocks
+                .iter()
+                .map(|base_block| {
+                    // This is an SMT holding updates from the `base_committed` block in mem.
+                    let updates = Self::gen_updates(&mut rng, &keys, base_block.updates.len());
+                    let proof_reader = Self::gen_proof_reader(&mut naive_base_smt, &updates);
+
+                    Block {
+                        smt: base_block
+                            .smt
+                            .batch_update(base_block.updates_flat_batch(), &base_block.proof_reader)
+                            .unwrap(),
+                        updates,
+                        proof_reader,
+                    }
+                })
+                .collect(),
+        };
+
+        Self {
+            base_empty,
+            base_committed,
+            base_uncommitted,
+        }
+    }
+
+    fn run(&self, c: &mut Criterion) {
+        self.base_empty.run(c);
+        self.base_committed.run(c);
+        self.base_uncommitted.run(c);
+    }
+
+    fn gen_updates(
+        rng: &mut StdRng,
+        keys: &[HashValue],
+        block_size: usize,
+    ) -> Vec<Vec<(HashValue, AccountStateBlob)>> {
+        std::iter::repeat_with(|| vec![Self::gen_update(rng, keys), Self::gen_update(rng, keys)])
+            .take(block_size)
+            .collect()
+    }
+
+    fn gen_update(rng: &mut StdRng, keys: &[HashValue]) -> (HashValue, AccountStateBlob) {
+        (*keys.iter().choose(rng).unwrap(), Self::gen_value(rng))
     }
-    group.finish();
+
+    fn gen_value(rng: &mut StdRng) -> AccountStateBlob {
+        rng.gen::<[u8; 32]>().to_vec().into()
```

Comment:
too small?

Reply:
It's planned that the global SMT leaves carry account "authenticator"s which are hash values.

Reply:
Actually.. that plus the chaining structure. I'l do 100bytes then.